### PR TITLE
Allow upload using binaries from build-path

### LINKED ISSUE #641: Upload build path is different than output build path
### Current behavior

```sh
arduino-cli --config-file /home/andrew/.config/arduino-cli/config.yaml compile --verbose --log-level debug --fqbn adafruit:nrf52:itsybitsy52840 --upload --port /dev/ttyACM1  --build-path /home/andrew/.tmp/arduino-build/bled --build-cache-path /home/andrew/.tmp/arduino-build.cache/bled  /home/andrew/Development/arduino/sketchbook/bled
```

The Adafruit nRF52840 core `platform.txt` contains a recipe to generate a bootable DFU .zip package specifying the output path as `{build.path}/{build.project_name}.zip`:
```sh
## Create dfu package zip file
recipe.objcopy.zip.pattern="{tools.nrfutil.cmd}" dfu genpkg --dev-type 0x0052 --sd-req {build.sd_fwid} --application "{build.path}/{build.project_name}.hex" "{build.path}/{build.project_name}.zip"
```

The upload utility recipe uses the exact same path `{build.path}/{build.project_name}.zip`:
```sh
tools.nrfutil.upload.pattern="{cmd}" {upload.verbose} dfu serial -pkg "{build.path}/{build.project_name}.zip" -p {serial.port} -b 115200 --singlebank
```

Yet when I execute the command pasted above (with a build path given at the command line), the outputs indicate two different paths are used:
```sh
...
Running recipe: recipe.objcopy.zip.pattern
adafruit-nrfutil dfu genpkg --dev-type 0x0052 --sd-req 0x00B6 --application /home/andrew/.tmp/arduino-build/bled/bled.ino.hex /home/andrew/.tmp/arduino-build/bled/bled.ino.zip
Zip created at /home/andrew/.tmp/arduino-build/bled/bled.ino.zip
...
Sketch uses 35652 bytes (4%) of program storage space. Maximum is 815104 bytes.
Global variables use 6612 bytes (2%) of dynamic memory, leaving 230956 bytes for local variables. Maximum is 237568 bytes.
No new serial port detected.
Usage: adafruit-nrfutil dfu serial [OPTIONS]

Error: Invalid value for "-pkg" / "--package": Path "/home/andrew/Development/arduino/sketchbook/bled/bled.adafruit.nrf52.itsybitsy52840.zip" does not exist.
Error during Upload: uploading error: exit status 2
```

If I modify either path pattern I referenced above in `platform.txt`, I can verify the output of both commands uses their respective modified paths, so I'm confident these are the recipes being used.

### Expected behavior

I would expect the path pattern `{build.path}/{build.project_name}.zip` to resolve to the same path in all recipes, and in particular that path should reflect what's given by command line option `--build-path`.

### Environment

- CLI version (output of `arduino-cli version`): `arduino-cli Version: 0.0.0-git Commit: 8759bf46`
- OS and platform: `Ubuntu 18.04.4 LTS`

### PR DESCRIPTION
**Pull request by cmaglie**
**Please check if the PR fulfills these requirements**
- [x] This PR replaces #655
- [x] The PR follows [our contributing guidelines](https://arduino.github.io/arduino-cli/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?**

New flags are added to the `upload` command:

```
  -n, --upload-from-build-path   Optional, use binaries produced in build path for uploading.
      --build-path string        Optional, specify build path (can be used with --upload-from-build-path).
```

With `-n, --upload-from-build-path` the CLI will upload binaries from the build path.
`--build-path` force a custom build-path instead of the build-path autogenerated by the CLI.

The new behaviour is now the default if chained compile + upload (`compile -u`) is used.

* **What is the current behavior?**
Currently:

1. `arduino-cli compile -u ...` will produce the binaries in the sketch folder will use them for the upload.
2. `arduino-cli compile -n -u ...` will not work (because the `-n/--dry-run` will not copy binaries in the sketch folder and consequently the upload will fail.

* **What is the new behavior?**
After the patch
1. `arduino-cli compile -u ...` will produce the binaries in the sketch folder but for the upload the binaries in the build path will be used for the upload. This should fix #641 at least for the chained build+upload.
2. `arduino-cli compile -n -u ...` will now  work, it will upload without creating the binaries in the sketch folder (because the `-n/--dry-run` will not copy binaries in the sketch folder but the upload will use the binaries from the build-path :+1: ).

* **Does this PR introduce a breaking change?**
It should not.

### CODE REVIEW

**Comment by rsora** [comment]
After going through your code and a bit of discussion with @cmaglie here my considerations:

**TL;DR: with #655 nad this PR, we are basically trying to solve a "Core configuration" problem, rather than a CLI problem**

This is our context:
- The CLI `upload` functionality is built upon the so called Java IDE "Export" functionality, that copies artifact to upload on the board in the sketch path
- The Java IDE instead, during the upload phase, uses the so called "sketch build path" as the place for upload artifact searching, and does not use the export path for upload purposes
Being the "Export" functionality a secondary Java IDE feature and not well documented at cores specification level, we a percentage of community cores (~20%) that do not implement correctly the "Export" (that means: mismatch between upload recipe file extension and export recipe file extension in most cases)

So, it seems that we have the following strategies to solve this issue in the CLI:
1. Ask to community cores to fix their export misconfiguration (and hope we get all the fixes merged in a reasonable amount of time) without changing the CLI behavior
2. Solve the problem merging this PR, maintaining the "Export" folder as the default upload artifact path, and giving the user flags for search in the build path for artifacts
3. Change the default CLI behavior, using the "build path" folder as the default upload artifact path. We can provide a `--export` flag for the compile to explicitly call for it
4. Ignore the "Export" configuration completely, and by default export all build artifacts generated in the build folder as <sketch_name>.ino.* without changing CLI behavior

Actually I believe that 4 could be the best strategy because:
- In 2-step compile + upload scenario, even if the compile uses different build path via flag, the upload command is not affected, as every artifact will be "exported" regardless of the core configuration
- In a combo compile --upload scenario ther is no need to pass anything to the "upload" invocation
- after having the CLI completely merged in the Java IDE, we can ask to community to "clean" their cores from export configurations, without worrying too much if te PRs will be merged in time or not.

@cmaglie WDYT?
